pipeline-model-api/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/DescriptorLookupCache.java

```diff
@@ -63,56 +64,71 @@ public DescriptorLookupCache() {
+        this.stepMap = CacheBuilder.newBuilder()
+                .expireAfterWrite(10, TimeUnit.MINUTES)
+                .build(new CacheLoader<String, StepDescriptor>() {
+                    @Override
+                    public StepDescriptor load(String key) throws Exception {
+                        for (StepDescriptor d : StepDescriptor.all()) {
+                            if (key.equals(d.getFunctionName())) {
+                                return d;
+                            }
+                        }
+
+                        return null;
+                    }
+                });
+
+        this.modelMap = CacheBuilder.newBuilder()
+                .expireAfterWrite(10, TimeUnit.MINUTES)
+                .build(new CacheLoader<String, DescribableModel<? extends Step>>() {
+                    @Override
+                    public DescribableModel<? extends Step> load(String key) throws Exception {
+                        StepDescriptor descriptor = lookupStepDescriptor(key);
+                        Class<? extends Step> c = (descriptor == null ? null : descriptor.clazz);
+                        return c != null ? new DescribableModel<>(c) : null;
+                    }
+                });
+
+        this.describableMap = CacheBuilder.newBuilder()
+                .expireAfterWrite(10, TimeUnit.MINUTES)
+                .build(new CacheLoader<String, Descriptor<? extends Describable>>() {
+                    @Override
+                    public Descriptor<? extends Describable> load(String key) throws Exception {
+                        try {
+                            return SymbolLookup.get().findDescriptor(Describable.class, key);
+                        } catch (NullPointerException e) {
+                            return null;
+                        }
+                    }
+                });
+
+        this.describableModelMap = CacheBuilder.newBuilder()
+                .expireAfterWrite(10, TimeUnit.MINUTES)
+                .build(new CacheLoader<String, DescribableModel<? extends Describable>>() {
+                    @Override
+                    public DescribableModel<? extends Describable> load(String key) throws Exception {
+                        final Descriptor<? extends Describable> function = lookupFunction(key);
+                        Class<? extends Describable> c = (function == null ? null : function.clazz);
+                        return c != null ? new DescribableModel<>(c) : null;
+                    }
+                });
     }
 
     public synchronized DescribableModel<? extends Step> modelForStep(String n) {
-        if (!modelMap.containsKey(n)) {
-            final StepDescriptor descriptor = lookupStepDescriptor(n);
-            Class<? extends Step> c = (descriptor == null ? null : descriptor.clazz);
-            modelMap.put(n, c != null ? new DescribableModel<>(c) : null);
-        }
-
-
-        return modelMap.get(n);
+        return modelMap.asMap().get(n);
```

Comment:
Are you sure this is what you want?
from the javadoc of `LoadingCache.asMap()`: **Note that although the view is modifiable, no method on the returned map will ever cause entries to be automatically loaded.**